File: backend/internal/models/automation.go
package models

import (
	"encoding/json"
	"time"
)

type Automation struct {
	ID          int              `json:"id" db:"id"`
	Name        string           `json:"name" db:"name"`
	Description string           `json:"description" db:"description"`
	TriggerType string           `json:"trigger_type" db:"trigger_type"`
	TriggerConfig json.RawMessage `json:"trigger_config" db:"trigger_config"`
	IsActive    bool             `json:"is_active" db:"is_active"`
	CreatedAt   time.Time        `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time        `json:"updated_at" db:"updated_at"`
	Steps       []AutomationStep `json:"steps,omitempty"`
}

type AutomationStep struct {
	ID           int             `json:"id" db:"id"`
	AutomationID int            `json:"automation_id" db:"automation_id"`
	StepOrder    int            `json:"step_order" db:"step_order"`
	StepType     string         `json:"step_type" db:"step_type"`
	Config       json.RawMessage `json:"config" db:"config"`
	DelayMinutes int            `json:"delay_minutes" db:"delay_minutes"`
	CreatedAt    time.Time      `json:"created_at" db:"created_at"`
}

type AutomationLog struct {
	ID           int             `json:"id" db:"id"`
	AutomationID int            `json:"automation_id" db:"automation_id"`
	StepID       *int64          `json:"step_id" db:"step_id"`
	SubscriberID *int64          `json:"subscriber_id" db:"subscriber_id"`
	Status       string          `json:"status" db:"status"`
	Details      json.RawMessage `json:"details" db:"details"`
	CreatedAt    time.Time       `json:"created_at" db:"created_at"`
}

type CreateAutomationRequest struct {
	Name          string                    `json:"name"`
	Description   string                    `json:"description"`
	TriggerType   string                    `json:"trigger_type"`
	TriggerConfig json.RawMessage           `json:"trigger_config"`
	Steps         []CreateAutomationStepReq `json:"steps"`
}

type CreateAutomationStepReq struct {
	StepOrder    int             `json:"step_order"`
	StepType     string          `json:"step_type"`
	Config       json.RawMessage `json:"config"`
	DelayMinutes int             `json:"delay_minutes"`
}

type UpdateAutomationRequest struct {
	Name          string                    `json:"name"`
	Description   string                    `json:"description"`
	TriggerType   string                    `json:"trigger_type"`
	TriggerConfig json.RawMessage           `json:"trigger_config"`
	IsActive      *bool                     `json:"is_active"`
	Steps         []CreateAutomationStepReq `json:"steps"`
}

File: frontend/src/services/automation.ts
import api from './api'

export interface AutomationData {
  id: number
  name: string
  description: string
  trigger_type: string
  trigger_config: Record<string, any>
  is_active: boolean
  steps: AutomationStepData[]
  created_at: string
  updated_at: string
}

export interface AutomationStepData {
  id: number
  automation_id: number
  step_order: number
  step_type: string // 'send_email' | 'wait' | 'condition'
  config: Record<string, any>
  delay_minutes: number
  created_at: string
}

export interface AutomationLogData {
  id: number
  automation_id: number
  step_id: number | null
  subscriber_id: number | null
  status: string
  details: Record<string, any>
  created_at: string
}

export interface CreateAutomationPayload {
  name: string
  description?: string
  trigger_type: string
  trigger_config?: Record<string, any>
  steps?: {
    step_order: number
    step_type: string
    config?: Record<string, any>
    delay_minutes?: number
  }[]
}

export interface UpdateAutomationPayload extends CreateAutomationPayload {
  is_active?: boolean
}

const automationService = {
  getAll: async (params?: { page?: number; per_page?: number }) => {
    const response = await api.get('/automations', { params })
    return response.data
  },

  getById: async (id: number) => {
    const response = await api.get(`/automations/${id}`)
    return response.data
  },

  create: async (data: CreateAutomationPayload) => {
    const response = await api.post('/automations', data)
    return response.data
  },

  update: async (id: number, data: UpdateAutomationPayload) => {
    const response = await api.put(`/automations/${id}`, data)
    return response.data
  },

  delete: async (id: number) => {
    const response = await api.delete(`/automations/${id}`)
    return response.data
  },

  toggleStatus: async (id: number) => {
    const response = await api.put(`/automations/${id}/toggle`)
    return response.data
  },

  getLogs: async (id: number, params?: { page?: number; per_page?: number }) => {
    const response = await api.get(`/automations/${id}/logs`, { params })
    return response.data
  }
}

export default automationService
